pkg/metrics/metrics.go

```diff
@@ -0,0 +1,186 @@
+	}, []string{"name", "namespace"})
+
+	// MachineSetStatusReadyReplicas Information of the Machinesets' status for ready replicas
+	MachineSetStatusReadyReplicas = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mapi_machine_set_status_ready_replicas",
+		Help: "Information of the mapi managed Machineset's status for ready replicas.",
+	}, []string{"name", "namespace"})
+
+	// MachineSetStatusReplicas Information of the mapi managed Machinesets' status for replicas.
+	MachineSetStatusReplicas = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mapi_machine_set_status_replicas",
+		Help: "Information of the mapi managed Machineset's status for replicas.",
+	}, []string{"name", "namespace"})
+
+	// ScrapeFailedCounter is a Prometheus metric, which counts errors during metrics collection.
+	ScrapeFailedCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "mapi_scrape_failure_total",
+		Help: "Total count of scrape failures.",
+	}, []string{"kind", "reason"})
+)
+
+func init() {
+	prometheus.MustRegister(ScrapeFailedCounter)
+	prometheus.MustRegister(MachineInfo)
+	prometheus.MustRegister(MachineSetInfo)
+	prometheus.MustRegister(MachineSetInfoSpecReplicas)
+	prometheus.MustRegister(MachineSetStatusAvailableReplicas)
+	prometheus.MustRegister(MachineSetStatusReadyReplicas)
+	prometheus.MustRegister(MachineSetStatusReplicas)
+}
+
+// MachineCollector is implementing prometheus.Collector interface.
+type MachineCollector struct {
+	machineLister    machinelisters.MachineLister
+	machineSetLister machinelisters.MachineSetLister
+	namespace        string
+}
+
+func NewMachineCollector(machineInformer machineinformers.MachineInformer, machinesetInformer machineinformers.MachineSetInformer, namespace string) *MachineCollector {
+	return &MachineCollector{
+		machineLister:    machineInformer.Lister(),
+		machineSetLister: machinesetInformer.Lister(),
+		namespace:        namespace,
+	}
+}
+
+// Collect is method required to implement the prometheus.Collector(prometheus/client_golang/prometheus/collector.go) interface.
+func (mc *MachineCollector) Collect(ch chan<- prometheus.Metric) {
+	mc.collectMachineMetrics(ch)
+	mc.collectMachineSetMetrics(ch)
+}
+
+// Describe implements the prometheus.Collector interface.
+func (mc MachineCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- MachineCountDesc
+	ch <- MachineSetCountDesc
+}
+
+// Collect implements the prometheus.Collector interface.
+func (mc MachineCollector) collectMachineMetrics(ch chan<- prometheus.Metric) {
+	machineList, err := mc.listMachines()
+	if err != nil {
+		ScrapeFailedCounter.With(prometheus.Labels{"kind": "Machine-count", "reason": err.Error()}).Inc()
+		return
+	}
+
+	for _, machine := range machineList {
+		mMeta := machine.ObjectMeta
+		typeMeta := machine.TypeMeta
+		mSpec := machine.Spec
+		providerid := ""
+		if mSpec.ProviderID != nil {
+			providerid = *mSpec.ProviderID
+		}
+
+		MachineInfo.With(prometheus.Labels{
```

Comment:
I'm not sure if it's relevant in this case but worth mentioning: once registered, the metric will continue to exist for the entire lifetime of the program even though the machine doesn't exist anymore.